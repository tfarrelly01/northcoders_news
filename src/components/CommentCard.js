import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import UserNotAuthorised from './UserNotAuthorised';
import DeleteComment from './DeleteComment';
import VoteUpOrDown from './VoteUpOrDown';

import { USERNAME } from '../config';

class CommentCard extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      displayMessage: false,
    };
    this.handleDelete = this.handleDelete.bind(this);
    this.toggleMessageDisplay = this.toggleMessageDisplay.bind(this);
    this.handleVoteUp = this.handleVoteUp.bind(this);
    this.handleVoteDown = this.handleVoteDown.bind(this);
  }

  handleDelete() {
    if (this.props.createdBy === USERNAME) this.props.deleteComment(this.props.id);
    else this.setState({ displayMessage: true });
  }

  toggleMessageDisplay() {
    this.setState({
      displayMessage: false,
    });
  }  

  handleVoteUp() {
    this.props.commentVote(this.props.id, 'up');
  }

  handleVoteDown() {
    this.props.commentVote(this.props.id, 'down');
  }

  render() {
    let contentToRender;
    if (!this.state.displayMessage) {
      contentToRender = (
        <div>
          <h6 className="subtitle is-6">
            By <strong>{this.props.name}</strong> 
            ({this.props.createdBy}): <strong>{moment(this.props.createdAt).fromNow()}</strong>
          </h6>
          <div className="content is-small">
            <p>{this.props.body}</p>
          </div>
        </div>
      );
    } else {
      contentToRender = (
        <UserNotAuthorised 
          displayMessage={this.state.displayMessage} 
          onClose={this.toggleMessageDisplay} 
        />
      );
    }
    return (
      <div className="container">
        <div className="box">
          <article className="media">
            <div className="media-left">
              <span>
                <VoteUpOrDown 
                  id={this.props.id}
                  avatarUrl={this.props.avatarUrl}
                  votes={this.props.votes}
                  handleVoteUp={this.handleVoteUp}
                  handleVoteDown={this.handleVoteDown}
                />
              </span>
            </div>

            <div className="media-content">
              <div className="content">
                {contentToRender}
              </div>
            </div>
            <DeleteComment handleDelete={this.handleDelete} />
          </article>
        </div>
      </div>
    );
  }
}

CommentCard.defaultProps = {
  body: '',
  id: '',
  votes: 0,
  createdAt: 0,
  avatarUrl: '',
  name: '',
  createdBy: '',
  commentVote: () => {},
  deleteComment: () => {},
};

CommentCard.propTypes = {
  body: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  createdAt: PropTypes.number.isRequired,
  avatarUrl: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  createdBy: PropTypes.string.isRequired,
  commentVote: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired
};

export default CommentCard;