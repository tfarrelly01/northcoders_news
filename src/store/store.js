import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import reducer from '../reducer/reducer';

const logger = createLogger();

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk, logger),
    typeof window === 'object' && typeof window.devToolsExtension !== 'undefined'
      ? window.devToolsExtension()
      : f => f
  )
);

export default store;