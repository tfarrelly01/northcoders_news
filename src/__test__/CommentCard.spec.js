import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../store/store';

import CommentCard from '../components/CommentCard';

describe('CommentCard Component', () => {
  test('CommentCard renders correctly', () => {
    const component = shallow(
      <Provider store={store}>
        <CommentCard />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});