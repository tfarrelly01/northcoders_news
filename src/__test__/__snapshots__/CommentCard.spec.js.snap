// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentCard Component CommentCard renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <CommentCard
    avatarUrl=""
    body=""
    commentVote={[Function]}
    createdAt={0}
    createdBy=""
    deleteComment={[Function]}
    id=""
    name=""
    votes={0}
  />,
  "nodes": Array [
    <CommentCard
      avatarUrl=""
      body=""
      commentVote={[Function]}
      createdAt={0}
      createdBy=""
      deleteComment={[Function]}
      id=""
      name=""
      votes={0}
    />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Provider
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
      >
        <CommentCard
          avatarUrl=""
          body=""
          commentVote={[Function]}
          createdAt={0}
          createdBy=""
          deleteComment={[Function]}
          id=""
          name=""
          votes={0}
        />
      </Provider>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Provider {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "children": <CommentCard
            avatarUrl=""
            body=""
            commentVote={[Function]}
            createdAt={0}
            createdBy=""
            deleteComment={[Function]}
            id=""
            name=""
            votes={0}
          />,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
        },
        "refs": Object {},
        "state": null,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <CommentCard
          avatarUrl=""
          body=""
          commentVote={[Function]}
          createdAt={0}
          createdBy=""
          deleteComment={[Function]}
          id=""
          name=""
          votes={0}
        />,
        "_debugID": 2,
        "_renderedOutput": <CommentCard
          avatarUrl=""
          body=""
          commentVote={[Function]}
          createdAt={0}
          createdBy=""
          deleteComment={[Function]}
          id=""
          name=""
          votes={0}
        />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <CommentCard
      avatarUrl=""
      body=""
      commentVote={[Function]}
      createdAt={0}
      createdBy=""
      deleteComment={[Function]}
      id=""
      name=""
      votes={0}
    />
  </Provider>,
}
`;
